static-metric/src/builder.rs

```diff
@@ -0,0 +1,227 @@
+        }
+    }
+
+    fn build_impl_from(&self) -> Tokens {
+        let struct_name = &self.struct_name;
+        let metric_vec_type = &self.metric_vec_type;
+
+        let prev_labels_ident: Vec<_> = (0..self.label_index)
+            .map(|i| Ident::from(format!("label_{}", i)))
+            .collect();
+        let body = self.build_impl_from_body(prev_labels_ident.clone());
+
+        quote!{
+            pub fn from(
+                #(
+                    #prev_labels_ident: &str,
+                )*
+                m: &#metric_vec_type
+            ) -> #struct_name {
+                #struct_name {
+                    #body
+                }
+            }
+        }
+    }
+
+    fn build_impl_from_body(&self, prev_labels_ident: Vec<Ident>) -> Tokens {
+        let member_type = &self.member_type;
+        let bodies: Vec<_> = self.label
+            .values
+            .iter()
+            .map(|value| {
+                let name = &value.name;
+                let value = &value.value;
+                if self.is_last_label {
+                    let current_label = &self.label.label_key;
+                    let prev_labels_str: Vec<_> = prev_labels_ident
+                        .iter()
+                        .enumerate()
+                        .map(|(i, _)| &self.metric.labels[i].label_key)
+                        .collect();
+                    let prev_labels_ident = prev_labels_ident.clone();
+                    quote!{
+                        #name: m.with(&{
+                            let mut coll = HashMap::new();
+                            #(
+                                coll.insert(#prev_labels_str, #prev_labels_ident);
+                            )*
+                            coll.insert(#current_label, #value);
+                            coll
+                        }),
+                    }
+                } else {
+                    let prev_labels_ident = prev_labels_ident.clone();
+                    quote!{
+                        #name: #member_type::from(
+                            #(
+                                #prev_labels_ident,
+                            )*
+                            #value,
+                            m,
+                        ),
+                    }
+                }
+            })
+            .collect();
+        quote!{
+            #(
+                #bodies
+            )*
+        }
+    }
+
+    fn build_impl_get(&self) -> Tokens {
+        let member_type = &self.member_type;
+        let values_str: Vec<_> = self.label.values.iter().map(|v| &v.value).collect();
+        let names_ident: Vec<_> = self.label.values.iter().map(|v| &v.name).collect();
+        quote!{
+            pub fn get(&self, value: &str) -> &#member_type {
+                match value {
+                    #(
+                        #values_str => &self.#names_ident,
+                    )*
+                    _ => panic!("unknown field `{}`", value),
```

Comment:
So caution should be taken when we `get`.

Reply:
Perhaps it should return an `Option<T>` instead?

Reply:
hmmmm.. Good idea. We can change to `Option` to leave the choice of panic to users.